tests/Unit/Domain/Test_ElementDistribution.cpp

```diff
@@ -0,0 +1,307 @@
+    for (size_t& val : number_of_clusters_per_proc) {
+      val = 0;
+    }
+    std::vector<bool> seen(gsl::at(proc_map, block).size());
+    // clang-tidy thinks this should be range-based, but that works weirdly with
+    // vectors of bools
+    /// NOLINTNEXTLINE(modernize-loop-convert)
+    for (size_t i = 0; i < seen.size(); ++i) {
+      seen.at(i) = false;
+    }
+    for (size_t start_element = 0;
+         start_element < gsl::at(proc_map, block).size(); ++start_element) {
+      if (not seen.at(start_element)) {
+        const size_t current_proc =
+            gsl::at(gsl::at(proc_map, block), start_element);
+        block_set_per_proc.at(current_proc).insert(block);
+        ++number_of_clusters_per_proc.at(current_proc);
+        seen.at(start_element) = true;
+        // perform a bredth-first search to get all elements in the cluster that
+        // share a proc and mark them seen
+        std::deque<size_t> next_elements;
+        auto insert_adjacent_unseen = [&next_elements, &current_proc, &proc_map,
+                                       &refinement_levels_by_block, &strides,
+                                       &seen,
+                                       &block](const size_t index) noexcept {
+          for (size_t i = 0; i < Dim; ++i) {
+            const size_t index_in_dim =
+                (index / gsl::at(strides, i)) %
+                two_to_the(
+                    gsl::at(gsl::at(refinement_levels_by_block, block), i));
+            if (index_in_dim > 0 and
+                not seen.at(index - gsl::at(strides, i)) and
+                gsl::at(gsl::at(proc_map, block),
+                        index - gsl::at(strides, i)) == current_proc) {
+              next_elements.push_back(index - gsl::at(strides, i));
+              seen.at(index - gsl::at(strides, i)) = true;
+            }
+            if (index_in_dim + 1 <
+                    two_to_the(gsl::at(
+                        gsl::at(refinement_levels_by_block, block), i)) and
+                not seen.at(index + gsl::at(strides, i)) and
+                gsl::at(gsl::at(proc_map, block),
+                        index + gsl::at(strides, i)) == current_proc) {
+              next_elements.push_back(index + gsl::at(strides, i));
+              seen.at(index + gsl::at(strides, i)) = true;
+            }
+          }
+        };
+        insert_adjacent_unseen(start_element);
+        // note that this loop cannot be converted to an STL iterator algorithm
+        // because `insert_adjacent_unseen` violates iterator stability via the
+        // insertions.
+        while (not next_elements.empty()) {
+          const size_t front = next_elements.front();
+          next_elements.pop_front();
+          insert_adjacent_unseen(front);
+        }
+      }
+    }
+    // verify that the distribution is well-clustered -- the Z-curve should
+    // ensure no more than 2 clusters for each core
```

Comment:
I understood that a cluster was the group of elements associated with a CPU... how can there be two on a core? is my understanding broken?

Reply:
Here I was considering a cluster as a collection of elements that shared a block and CPU that were adjacent to one another (any cluster member shares at least one face with another member of the same cluster).
The point I'm making here is that the Z-order distribution performs diagonal 'jumps', and the ordering ensures that there are either one or two such clusters for each core for each block that core participates in. This is to be contrasted with a Hilbert curve, for which the construction would ensure exactly one cluster under the same definitions.